### PR TITLE
SCMOD-4128: Fix erroneous email detection

### PR DESCRIPTION
**Pull request by morvinshah4**
https://jira.autonomy.com/browse/SCMOD-4128

### CODE REVIEW

**Comment by morvinshah4** [comment]
https://jira.autonomy.com/browse/SCMOD-4128

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java [hunk 1]
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
...
+                emailElements.get(i).detach();
+            }
+            if (header.getContentSize() == 0) {
+                String currentBody = ((Element) emailElements.get(i).getContent().get(1)).getText();
+                String previousBody = ((Element) emailElements.get(i-1).getContent().get(1)).getText();
+                String newBody = previousBody + currentBody;
+                ((Element) emailElements.get(i-1).getContent().get(1)).setText(newBody);
+                emailElements.get(i).detach();
+            }
+        }
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java]: A lot of these variables can be final.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java [hunk 2]
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
             }
         }
         LOG.info("Markup of Headers and Body complete");
+
+        // Fix for SCMOD-4128 to untag and detach erroneous emails which have blank headers.
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java]: Don't mention the Jira number in the comments.  Just describe the behavior.  Describe for example why you are starting at index 1 rather than index 0.  A casual reader may think you've made a mistake with something unusual like that.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java [hunk 3]
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
...
+        for (int i = 1; i < emailElements.size(); i++) {
+
+            Element header = (Element) emailElements.get(i).getContent().get(0);
+            Element body = (Element) emailElements.get(i).getContent().get(1);
+
+            if (header.getContentSize() == 0 && body.getContentSize() == 0)  {
+                emailElements.get(i).detach();
+            }
+            if (header.getContentSize() == 0) {
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java]: Is there a missing `else` here?

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java [hunk 4]
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
...
+                String currentBody = ((Element) emailElements.get(i).getContent().get(1)).getText();
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java]: Why is this not just `body.getText()`?

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java (hunk 1 above)
Comment (dermot-hardy) [reply worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java]: Agreed - they should all be `final`.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
You need to some tests for this functionality.

**Comment by Anthony-McGreevy** [comment]
There seems to be a lot of whitespace changes in the test class.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ */
+package com.github.cafdataprocessing.worker.markup.core;
+
+import org.jdom2.Element;
+
+import java.util.List;
+
+/**
+ * Created by Morvin Shah on 6/14/2018.
+ */
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: remove this comment

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+public final class EmailSquash {
+
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: maybe try and think of a better name that conveys what the class is being used for

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    private EmailSquash(){}
+    public static void untagFalseEmails(final List<Element> emailElements)
+
+    {
+        for (int i = emailElements.size() - 1; i > 0; i--) {
+            final Element header = (Element) emailElements.get(i).getContent().get(0);
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: add some comments to the method code and a javadoc to the method.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java [hunk 8]
```diff
@@ -166,7 +167,7 @@ else if (line.contains(": ")) {
                 else if (line.contains(":*")) {
                     bodyIndex++;
                     addStandardisedHeader(nattyParser, headersElement, lines, line, ":\\*");
-                }else {
+                } else {
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java]: Remove whitespace changes from the commit. These are good to get in but they should be in their own commit, that way if we back out a feature we don't also lose good formatting changes.

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 9]
```diff
@@ -278,6 +278,20 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
         assertTrue(TestUtility.compareHeaderElements(docForComparison, xmlDocument));
     }
 
+   @Test
+    public void testEmailUntag() throws IOException, JDOMException
+    {
+       Document xmlDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntag.xml");
+       MarkupHeadersAndBody.markUpHeadersAndBody(xmlDocument, emailHeaderMappings, condensedHeaderMultiLangMappings);
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: Make these variables final

**Review by Anthony-McGreevy — DISMISSED** [review]

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 10]
```diff
@@ -278,6 +278,20 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
         assertTrue(TestUtility.compareHeaderElements(docForComparison, xmlDocument));
     }
 
+   @Test
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: Add a comment to say what this test is actually testing.

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 11]
```diff
@@ -278,6 +278,20 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
         assertTrue(TestUtility.compareHeaderElements(docForComparison, xmlDocument));
     }
 
+    @Test
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: All the other tests have comments describing what they do.  This one you've added doesn't.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.List;
+
+public final class EmailSquash
+{
+    private EmailSquash()
+    {
+    }
+
+    public static void untagFalseEmails(final List<Element> emailElements)
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: The class or method could do with some JavaDoc to describe what it is trying to achieve.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        for (int i = emailElements.size() - 1; i > 0; i--) {
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: Add a comment to explain why this is `i > 0` rather than `i >= 0`.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
- So far as I can see the testcase that you've added doesn't actually test your code.  It doesn't contain any e-mails that don't have headers.
- On top of that you should have more that one testcase to test the different branches of your code and different placements of e-mails that need to be removed.

File: worker-markup-core/src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
...>\\rahul-agent03\testdata\1\4EDB8C21-89FF-11E5-9298-005056C00008 (26).eml</FILEPATH><FILE_EXT>eml</FILE_EXT><FILE_NAME>4EDB8C21-89FF-11E5-9298-005056C00008 (26).eml</FILE_NAME><FILE_SIZE>4155</FILE_SIZE><FROM>"Dan J Hyvl" &lt;Dan J Hyvl&gt;</FROM><HAS_ATTACHMENTS>false</HAS_ATTACHMENTS><ID>1022</ID><ID_PATH>/1022</ID_PATH><IS_FAMILY_ORIGIN>true</IS_FAMILY_ORIGIN><IS_OLE>false</IS_OLE><IS_ROOT>true</IS_ROOT><JOB_ID>1148</JOB_ID><JOB_RUN_ID>1313</JOB_RUN_ID><MESSAGE_ID>&lt;HADDMVBYL4NMUHYRWKEW1G0NDSGJJT02A@zlsvr22&gt;</MESSAGE_ID><MIMETYPE>message/rfc822</MIMETYPE><REPOSITORY_ID>1147</REPOSITORY_ID><REPOSITORY_SUB_TYPE>FILESYSTEMWIN</REPOSITORY_SUB_TYPE><REPOSITORY_TYPE>FILESYSTEM</REPOSITORY_TYPE><ROOT_ID>1022</ROOT_ID><SCAN_REFERENCE>718b1ba0df2840a95988a02bcba7e4fe|1527724549|e5609295d75cc8c9909588d81f7bc391|00000000000000000000000000000000|4155</SCAN_REFERENCE><SOURCE_REFERENCE>\\rahul-agent03\testdata\1\4EDB8C21-89FF-11E5-9298-005056C00008 (26).eml|1527724549|e5609295d75cc8c9909588d81f7bc391</SOURCE_REFERENCE><TO>"Pfister Christian W." &lt;CWPfister@cps-satx.com&gt;</TO><priority>0</priority><sent>2001-01-30T17:37:00Z</sent><subject>RE: Enfolio Contract with CPS</subject></root>
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml]: If this test is passing then the logic must not work since I can clearly see an e-mail with no headers and no body:

    <email><headers></headers><body></body></email>

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: worker-markup-core/src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml (hunk 14 above)
Comment (morvinshah4) [reply worker-markup-core/src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml]: Dermot I added the empty header and body after our call to test the other if condition

Comment (dermot-hardy) [reply worker-markup-core/src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml]: This is the expectation file.  It is being tested against the output to confirm correct behavior.  To test the other `if` condition you'd need to adjust the input file, or better still add a separate test for it.

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 15]
```diff
@@ -284,15 +284,14 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
...
-        MarkupHeadersAndBody.markUpHeadersAndBody(xmlDocument, emailHeaderMappings, condensedHeaderMultiLangMappings);
-        Document docForComparison = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml");
+        Document expectedDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBody.xml");
+        MarkupHeadersAndBody.markUpHeadersAndBody(expectedDocument, emailHeaderMappings, condensedHeaderMultiLangMappings);
+        Document actualDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml");
 
-        String docMarkedupValue = xmlDocument.getRootElement().getValue();
-        String docForComparisonValue = docForComparison.getRootElement().getValue();
+        String docMarkedupValue = expectedDocument.getRootElement().getValue();
+        String docForComparisonValue = actualDocument.getRootElement().getValue();
```
Comment (Anthony-McGreevy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: All of the variables in these test methods can be final.

**Comment by Anthony-McGreevy** [comment]
@morvinshah4 Is there a dev build of this where is can see these tests passing?

**Comment by morvinshah4** [comment]
http://sou-jenkins2.hpeswlab.net/job/CAFDataProcessing/view/Developer/job/CAFDataProcessing~worker-markup~SCMOD-4128~CI/

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
What we discussed earlier about the xml strings being compared hasn't been implemented.  When we talked about it you knew the correct functions to call but they are not being called here.

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 16]
```diff
@@ -278,6 +279,38 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
...
 
+    /*
+     * Test for Untaging emails which have empty headers and empty body.
+     */
+    @Test
+    public void testEmailUntagWithEmptyHeadersEmptyBody() throws IOException, JDOMException
+    {
+        final Document expectedDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBody.xml");
+        MarkupHeadersAndBody.markUpHeadersAndBody(expectedDocument, emailHeaderMappings, condensedHeaderMultiLangMappings);
+        final Document actualDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml");
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: This is the expected file - not the actual

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 17]
```diff
@@ -278,6 +279,38 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
...
+
+        final String docMarkedupValue = new XMLOutputter().outputString(actualDocument);
+        final String docForComparisonValue = new XMLOutputter().outputString(expectedDocument);
+
+        assertTrue(docForComparisonValue.equals(docMarkedupValue));
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: We should be using `assertEquals` for a case like this.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
Looking at the testcases I can see that there may be `divider` elements between the `email` elements.  Looking at the code I can see that it doesn't take this into account.  The code looks OK if the invalid e-mail being squashed is preceded by a valid e-mail, but if it is preceded by a divider then it fails.

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 18]
```diff
@@ -278,6 +279,65 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
         assertTrue(TestUtility.compareHeaderElements(docForComparison, xmlDocument));
     }
 
+    /*
+     * Test for Untaging emails which have empty headers and empty body.
+     */
+   @Test
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: `@Test` annotation not aligned like the rest of them.

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 19]
```diff
@@ -278,6 +279,65 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
...
+        final String actualDocumentXml = new XMLOutputter().outputString(xmlDocument);
+        final String expectedDocumentXml = new XMLOutputter().outputString(expectedDocument);
+
+        assertEquals(expectedDocumentXml, actualDocumentXml);
+    }
+
+    /*
+     * Test for Untaging emails which have empty headers and non empty body.
+     */
+   @Test
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: `@Test` annotation not aligned like the rest of them.

File: worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java [hunk 20]
```diff
@@ -278,6 +279,65 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
...
+        assertEquals(preMarkupValue, postMarkupValue);
+
+        final Document expectedDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagHavingDividerWithEmptyHeadersEmptyBodyMarkedup.xml");
+
+        final String actualDocumentXml = new XMLOutputter().outputString(xmlDocument);
+        final String expectedDocumentXml = new XMLOutputter().outputString(expectedDocument);
+
+        assertEquals(expectedDocumentXml, actualDocumentXml);
+    }
     /**
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java]: Leave a blank line rather than running straight into the next function.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * Untags emails with blank headers except the parent email.
+     *
+     * @param parentElement The number of emails from which the false ones are to untagged
+     */
+    public static void untagFalseEmails(final Element parentElement) {
+        final List<Element> emailElements = parentElement.getChildren();
+
+        // The loop starts with 1 because the we do not have header information of the parent email
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: This comment makes no sense - the loop doesn't start with 1.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+        for (int i = emailElements.size() - 1; i > 0; i--) {
+            final Element previousElement = emailElements.get(i - 1);
+            final Element currentElement = emailElements.get(i);
+            INNERLOOP:
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: Why do we need to use a label here?  Also, what inner loop are you referring to here?  You appear to be labeling an `if` statement.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+            if ((currentElement.getName().equals("email"))) {
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: There appear to be extra unnecessary sets of brackets in this line.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+                final Element headersElement = (Element) currentElement.getContent().get(0);
+                final Element bodyElement = (Element) currentElement.getContent().get(1);
+                if ((previousElement.getName().equals("divider"))) {
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: There appear to be extra unnecessary sets of brackets in this line.

File: worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+                    if (headersElement.getContentSize() == 0 && (bodyElement.getContentSize() == 0  ||
+                            bodyElement.getContent().get(0).getValue().equals(""))) {
```
Comment (dermot-hardy) [inline worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java]: - Add an additional check that there is only one content.  At the minute it appears to be an implicit assumption (although most likely a valid one).
- Use `.isEmpty()` rather than `.equals("")`

File: worker-markup-core/src/test/resources/xml/EmailUntagHavingDividerWithEmptyHeadersEmptyBody.xml [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
...ian W."}</ACCOUNTS><ADDRESS_DISPLAY_FROM>"Dan J Hyvl" &lt;Dan J Hyvl&gt;</ADDRESS_DISPLAY_FROM><ADDRESS_DISPLAY_TO>Pfister Christian W.</ADDRESS_DISPLAY_TO><ADDRESS_FROM>dan j hyvl &lt;dan j hyvl&gt;</ADDRESS_FROM><ADDRESS_TO>cwpfister@cps-satx.com</ADDRESS_TO><ARCHIVE_ID>50fcf7df-3f9c-4ebc-86fb-c09669a01158/cbf14d0c-ac87-45c7-8d47-a36d6cbb8e00.caf</ARCHIVE_ID><BATCH_ID>71</BATCH_ID><BINARY_HASH_SHA1>464c0fb8695eefb048e59d2b57673bffef757c2e</BINARY_HASH_SHA1><CAF_INDEXING_TOKEN>{"INDEX_ROUTE":"1022","INDEX_NAME":"morvinnew_item_1"}</CAF_INDEXING_TOKEN><CAF_MAIL_CONVERSATION_INDEX>AcnSAfKIY9ehzkbQT9eazV9TYssN5A==</CAF_MAIL_CONVERSATION_INDEX><CAF_MAIL_CONVERSATION_INDEX_PARSED>{"headerDate":"2009-05-11T06:30:16.256Z","guid":"63d7a1ce-46d0-4fd7-9acd-5f5362cb0de4","childMessages":[]}</CAF_MAIL_CONVERSATION_INDEX_PARSED><CAF_WORKFLOW_ACTION>1</CAF_WORKFLOW_ACTION><CAF_WORKFLOW_ID>1</CAF_WORKFLOW_ID><CAF_WORKFLOW_OUTPUT_PARTIAL_REFERENCE>50fcf7df-3f9c-4ebc-86fb-c09669a01158</CAF_WORKFLOW_OUTPUT_PARTIAL_REFERENCE><CAF_WORKFLOW_PROJECT_ID>1</CAF_WORKFLOW_PROJECT_ID><CAF_WORKFLOW_TENANT_ID>morvinnew</CAF_WORKFLOW_TENANT_ID><COLLECTION_STATUS>COLLECTED</COLLECTION_STATUS><CONTENT><email>Yes
```
Comment (dermot-hardy) [inline worker-markup-core/src/test/resources/xml/EmailUntagHavingDividerWithEmptyHeadersEmptyBody.xml]: The rest of the testcase files only have the `<CONTENT>` element and that is all that you require as well.  Drop all the redundant elements.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: worker-markup-core/src/test/resources/xml/EmailUntagHavingDividerWithEmptyHeadersEmptyBody.xml (hunk 26 above)
Comment (dermot-hardy) [reply worker-markup-core/src/test/resources/xml/EmailUntagHavingDividerWithEmptyHeadersEmptyBody.xml]: This comment still hasn't been addressed.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Review by dermot-hardy — APPROVED** [review]
